src/main/java/frc/robot/subsystems/ShooterSubsystem.java

```diff
@@ -0,0 +1,210 @@
+        new SetShooter(
+            calculateShooterSpeeds(0),
+            0)); // Replace distance and exit code with LimelightGetDistance() and
+    // CheckLimelightStatus() respectively
+  }
+
+  // Idle shooter command; for default command purposes
+  public Command shooterIdle() {
+    return run(() -> {
+          setShooterSpeeds(speakerConfigIdle);
+        })
+        .withName("Idle shooter command");
+  }
+
+  // Checks if shooter is ready.
+  public void isShooterReady(SpeakerConfig targetSpeeds) {
+    double targetSpeedTop = targetSpeeds.getSpeedTopInRPM();
+    double targetSpeedBottom = targetSpeeds.getSpeedBottomInRPM();
+
+    topMotorTargetSpeedEntry.log(targetSpeedTop);
+    bottomMotorTargetSpeedEntry.log(targetSpeedBottom);
+
+    double minSpeedToleranceTop =
+        targetSpeedTop * (1 - Constants.ShooterConstants.SHOOTER_SPEED_TOLERANCE);
+    double maxSpeedToleranceTop =
+        targetSpeedTop * (1 + Constants.ShooterConstants.SHOOTER_SPEED_TOLERANCE);
+    double minSpeedToleranceBottom =
+        targetSpeedBottom * (1 - Constants.ShooterConstants.SHOOTER_SPEED_TOLERANCE);
+    double maxSpeedToleranceBottom =
+        targetSpeedBottom * (1 + Constants.ShooterConstants.SHOOTER_SPEED_TOLERANCE);
+
+    double speedTop = shooterTopMotor.getEncoder().getVelocity();
+    double speedBottom = shooterBottomMotor.getEncoder().getVelocity();
+
+    topMotorSpeedEntry.log(speedTop);
+    bottomMotorSpeedEntry.log(speedBottom);
+
+    if ((minSpeedToleranceTop < speedTop && speedTop < maxSpeedToleranceTop)
+        && (minSpeedToleranceBottom < speedBottom && speedBottom < maxSpeedToleranceBottom)
+            == true) {
+      // Request note from feeder.
+    }
+  }
+
+  // Speed in RPM. Top is index 0, bottom is index 1.
+  public SpeakerConfig setShooterSpeeds(SpeakerConfig speeds) {
+    SpeakerConfig targetSpeeds = speeds;
+
+    shooterTopMotor
+        .getPIDController()
+        .setReference(speeds.getSpeedTopInRPM(), CANSparkBase.ControlType.kVelocity);
+    shooterBottomMotor
+        .getPIDController()
+        .setReference(speeds.getSpeedBottomInRPM(), CANSparkBase.ControlType.kVelocity);
+
+    return targetSpeeds;
+  }
+
+  // Top is index 0, bottom is index 1.
+  public static SpeakerConfig calculateShooterSpeeds(double distance) {
+    SpeakerConfig speeds;
+    Double topSpeed = 0d;
+    Double bottomSpeed = 0d;
+    double distanceProportion;
+
+    // If distance below minimum, set speed to minimum.
+    if (distance < speakerConfigList[0].getDistanceInInches()) {
+      topSpeed = speakerConfigList[0].getSpeedTopInRPM();
+      bottomSpeed = speakerConfigList[0].getSpeedBottomInRPM();
+      speeds = new SpeakerConfig(distance, topSpeed, bottomSpeed);
+      return speeds;
+    }
+    // A linear search which determines which points the input distance falls between. May be
+    // converted to a binary search if there are many points
+    for (int i = 0; i < speakerConfigList.length - 1; i++) {
```

Comment:
You should add a comment calling out the fact that you're looping through all but the last element here. You might even consider changing it to `i <= length - 2` to make it more obvious that you're doing an un-conventional loop.

Reply:
Resolved in ccf2513.